### ISSUE TITLE
Basics

### ISSUE THREAD

**Issue by joobei** [KacheFlowe/Swift-AI#1]
Basel I had a hard time understanding the vanishing gradient problem so I decided to take Geoffrey Hinton's course on coursera. So it's going to take a while but it's proving to be very helpful for my understanding... 

![screen shot 2016-03-16 at 16 54 03](https://cloud.githubusercontent.com/assets/2638631/13819618/79a196ea-eb9a-11e5-9634-4dc611232f74.png)

**Comment by Jxrgxn** [comment]
Cool beans. I'm the type to dive in and get messy, but I might check this out...  I'll try to check in some code when I get my head above water. I've read as much as I can from WildML. I'll send you a great NN guide I received via email as well.

**Comment by joobei** [comment]
I tried Jorden Hill's birdbrain and at least the training the RNN seems to be working. It's MIT licensed so I can't see why we can't use his source as is.

**Comment by Jxrgxn** [comment]
But where's the fun in that?

**Comment by Jxrgxn** [comment]
@joobei I think we can definitely use the code as a guide or base. My hope however is for us to build something through our own efforts. I've been consumed by a project recently and so I haven't had time to work but I should now.

**Comment by Jxrgxn** [comment]
http://neuralnetworksanddeeplearning.com/chap5.html

**Comment by Jxrgxn** [comment]
http://arxiv.org/abs/1507.06947

**Comment by Jxrgxn** [comment]
I'm planning out my code today.

**Comment by joobei** [comment]
I ,on the other hand, being more pressed with the practical application of the RNN am still fighting it out with Birdbrain. That being said I think the lessons learned from me testing birdbrain will come in handy to us here too.

**Comment by Jxrgxn** [comment]
I'm assuming you're attempting to apply the model to the research we discussed?

**Comment by joobei** [comment]
Yes but I'm stuck at trying to train it. I'll report if I manage to get it working.

**Comment by Jxrgxn** [comment]
In the link I posted... Chapter 6 goes a great way to training. The Cornell Paper I posted also covers training.

**Comment by Jxrgxn** [comment]
@joobei what sort of extension are you using on your model, joobei? LSTM? That's what I'm going with.

**Comment by joobei** [comment]
@KacheFlowe I haven't built my own model. I'm just trying out birdbrain's vanilla RNN model for now. He also has an LSTM there if you want to take a look. I had trouble training the RNN so I am currently trying to build some test data set that I KNOW will work for an RNN. Because at this moment I don't know if the problem is my training data or birdbrain's RNN.

Would be nice to have a basic set of data that we KNOW works ...

**Comment by Jxrgxn** [comment]
Actually, I might have a solution to that... Hold on a moment

**Comment by Jxrgxn** [comment]
P.S. There's a REALLY great guide on LSTM: http://colah.github.io/posts/2015-08-Understanding-LSTMs/

**Comment by Jxrgxn** [comment]
@joobei Got it! 15,000 Reddit Comments courtesy of Google BigQuery: https://bigquery.cloud.google.com/table/fh-bigquery:reddit_comments.2015_08

**Comment by joobei** [comment]
Yes. In fact I was thinking of using the same dataset as that RNN tutorial I linked you. He uses data from http://www.nltk.org/

**Comment by joobei** [comment]
Here's their data set: http://www.nltk.org/nltk_data/

**Comment by Jxrgxn** [comment]
Do it!

**Comment by Jxrgxn** [comment]
@joobei You see, what happens @joobei at around 4 in the morning without a day of sleep, you make stupid decisions like "I need to make another repo to merge this code. Into my other code. ...What do you mean I can make a branch? ...Shut up, brain." So that's what happened.

**Comment by joobei** [comment]
update: I've tried and tried with the Birdbrain RNN but it seems to be buggy, so I'm possibly now going to look at trying to build my own RNN.

**Comment by Jxrgxn** [comment]
I'm already building my own. We can combine efforts. Right now I'm working
on a tokenizer.

**Comment by Jxrgxn** [comment]
@joobei remind me what you're parsing for?

**Comment by joobei** [comment]
@KacheFlowe This is what I'm parsing for:
[stats.stackExchange question](http://stats.stackexchange.com/questions/203272/how-can-i-find-patterns-or-irregularities-in-a-path-rnn)

**Comment by joobei** [comment]
This is what I think I'll be feeding the algorithm:

![rnn-data](https://cloud.githubusercontent.com/assets/2638631/14072077/ceec85b4-f4c4-11e5-8c9a-c7b427055c8e.png)

**Comment by joobei** [comment]
ps. I have decided to try Alex Grave's LSTM RNN before I try to code my own. His RNN was built specifically for connected handwriting recognition so it could work for me. It's in C++ though.

[github rnnlib](https://github.com/szcom/rnnlib)

**Comment by Jxrgxn** [comment]
This is great! Between the two of us--you can focus on handwriting and I can focus on actual translation

**Comment by Jxrgxn** [comment]
Will you be writing the auxiliary python scripts? seems complicated.

**Comment by Jxrgxn** [comment]
Hey @joobei what do you envision the ideal number of layers?

**Comment by joobei** [comment]
@KacheFlowe no idea what an appropriate number of layers is for my application. It really depends I think. Too many layers though and you could end up overfitting on some types of neural networks.
